File: AI_Engine_Development/AIE/Feature_Tutorials/26-gentle_intro_to_aie_kernel_programming/unit_tests/mtxvec/julia/mtxvec.jl
#====================================================================
  Copyright © 2025 Advanced Micro Devices, Inc. All rights reserved.
  SPDX-License-Identifier: MIT
====================================================================#

debug = true    # comment out to disable debug messages

include("../../../contrived/julia/utilities.jl")

# !!! User Parameters !!! #
# !!! Make sure that they match "params.h" !!!
mrows = 4       # no. of rows in input matrices
mcols = 4       # no. of columns in input matrices
burst_count = 8 # number of "samples" to process per function invocation
iter_count = 2  # no. of function invocations

# !!! End User Parameters !!! #

# !!! Derived Parameters !!! #
N = burst_count * iter_count    # total number of "samples" to generate
vlen = mcols * 2				# input vector length

# !!! End Derived Parameters !!! #

using Random
Random.seed!(11)    # set seed for repeatability

# create input matrices
M_ul = rand(Int16, (mrows, mcols, N))	# upper left matrix
M_ll = rand(Int16, (mrows, mcols, N))	# lower left matrix
M_ur = rand(Int16, (mrows, mcols, N))	# upper right matrix
M_lr = rand(Int16, (mrows, mcols, N))	# lower right matrix

# create input vectors
vi = rand(Int16, (vlen, N))

# reserve space
M = Array{Int32, 3}(undef, mrows * 2, mcols * 2, N)
vo = Array{Int32, 2}(undef, vlen, N)

# concatenate input matrices and calculate product
for i = 1 : N
	ul = M_ul[:, :, i]
	ll = M_ll[:, :, i]
	ur = M_ur[:, :, i]
	lr = M_lr[:, :, i]
	M[:, :, i] = vcat(hcat(ul, ur), hcat(ll, lr))	# concatenate matrices
	vo[:, i] = M[:, :, i] * vi[:, i]				# calculate product
end # for i = 1 : N

# write results
# !!! make sure that the PLIO width used matches that in src/graph.hpp !!!
arr2file(M_ul, true, 64, "ul.dat")
arr2file(M_ll, true, 64, "ll.dat")
arr2file(M_ur, true, 64, "ur.dat")
arr2file(M_lr, true, 64, "lr.dat")
arr2file(vi, false, 64, "vi.dat")
arr2file(vo, false, 64, "vo.dat")

run(`mv ul.dat ../data/.`)	# move stimulus files to data directory
run(`mv ll.dat ../data/.`)
run(`mv ur.dat ../data/.`)
run(`mv lr.dat ../data/.`)
run(`mv vi.dat ../data/.`)

File: AI_Engine_Development/AIE/Feature_Tutorials/26-gentle_intro_to_aie_kernel_programming/unit_tests/mtxvec/src/mtxvec.cpp
/********************************************************************
* Copyright © 2025 Advanced Micro Devices, Inc. All rights reserved.
* SPDX-License-Identifier: MIT
********************************************************************/

#include "aie_api/aie.hpp"
#include "aie_api/aie_adf.hpp"

#include "mtxvec.hpp"

//#define KDEBUG	// comment out to suppress debug messages

#ifdef KDEBUG
	#include "aie_api/utils.hpp"    // "pretty-print" contents of vector registers
#endif // KDEBUG

// concatenates 4 sub-matrices into a larger matrix
// then multiplies the resulting matrix with an input vector
template<typename Ti, typename Tacc, unsigned mrows, unsigned mcols, unsigned burst_count>
void mtxvec(
    adf::input_buffer<Ti, adf::extents<burst_count * mrows * mcols>> &UL_in,  // input upper left matrix
    adf::input_buffer<Ti, adf::extents<burst_count * mrows * mcols>> &LL_in,  // input lower left matrix
    adf::input_buffer<Ti, adf::extents<burst_count * mrows * mcols>> &UR_in,  // input upper right matrix
    adf::input_buffer<Ti, adf::extents<burst_count * mrows * mcols>> &LR_in,  // input lower right matrix
    input_stream<Ti> *istrm,        // input vector
    output_cascade<Tacc> *ocstrm    // output accumulator cascade
) {

    // initialize iterators for input submatrices
    auto ul_ptr = aie::begin_vector<mrows * mcols>(UL_in);
    auto ll_ptr = aie::begin_vector<mrows * mcols>(LL_in);
    auto ur_ptr = aie::begin_vector<mrows * mcols>(UR_in);
    auto lr_ptr = aie::begin_vector<mrows * mcols>(LR_in);

    using MMUL = aie::mmul<(mrows * 2), (mcols * 2), 1, Ti, Ti, Tacc>;    // alias for matrix-vector multiplication class
    MMUL mvmul; // instantiate matrix-vector multiplication class

    #ifdef KDEBUG
        printf("\n");
        printf("Running mtxvec() with %d sample%s...\n", burst_count, (burst_count > 1) ? "s": "");
    #endif // KDEBUG

    for (auto i = 0u; i < burst_count; i++) {

        #ifdef KDEBUG
            auto p_ul = ul_ptr;
            auto p_ur = ur_ptr;
            auto p_ll = ll_ptr;
            auto p_lr = lr_ptr;
        #endif // KDEBUG
        
        // concatenate sub-matrices
        auto [t1, t2] = aie::interleave_zip(*ul_ptr++, *ur_ptr++, mcols);
        auto top = aie::concat(t1, t2);
        auto [b1, b2] = aie::interleave_zip(*ll_ptr++, *lr_ptr++, mcols);
        auto bottom = aie::concat(b1, b2);
        auto m = aie::concat(top, bottom);

        auto v = readincr_v<mcols * 2>(istrm);  // form vector from stream
        mvmul.mul(m, v);                        // perform matrix-vector multiplication
        auto acc = mvmul.to_accum();            // put results in accumulator register
        writeincr(ocstrm, acc);                 // write accumulator to cascade stream

        #ifdef KDEBUG 
            printf("  i = %d:\n", i);
            aie::print_matrix(*p_ul++, mcols, "    a = ");
            printf("\n");
            aie::print_matrix(*p_ll++, mcols, "    b = ");
            printf("\n");
            aie::print_matrix(*p_ur++, mcols, "    d = ");
            printf("\n");
            aie::print_matrix(*p_lr++, mcols, "    e = ");
            printf("\n");
            aie::print_matrix(m, 2 * mcols, "    m = ");
            printf("\n");
            aie::print(v, true, "    v = ");
            printf("\n");
            auto p = mvmul.template to_vector<int32>();
            aie::print(p, true, "    p = ");
            printf("\n");
        #endif // KDEBUG
        
    } // end for (auto i = 0u; i < burst_count; i++)

} // end mtxvec()
